proposals/2020-07-01_aws-kms-keyring-redesign/change.md

```diff
@@ -0,0 +1,327 @@
+   3. MAY be configured with a list of string [grant tokens](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#grant-tokens)
+      to be included in all AWS KMS calls
+   4. MUST configure a multi-keyring of AWS KMS symmetric keyrings...
+      1. with a single AWS KMS symmetric keyring for each configured key name
+      2. with each AWS KMS symmetric keyring configured with the configured grant tokens, if applicable
+      3. where the order of the child keyrings matches the order of the configured key names
+2. **AWS KMS symmetric multi-region discovery keyring**
+   1. MUST be configured with a list of AWS regions
+   2. MAY be configured with an AWS account ID
+   3. MAY be configured with a list of string [grant tokens](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#grant-tokens)
+      to be included in all AWS KMS calls
+   4. MUST configure a multi-keyring of AWS KMS symmetric region discovery keyrings...
+      1. with a single AWS KMS symmetric region discovery keyring for each configured AWS region
+      2. with each AWS KMS symmetric region discovery keyring configured with the configured AWS account ID, if applicable
+      3. with each AWS KMS symmetric region discovery keyring configured with the configured grant tokens, if applicable
+      4. where the order of the child keyrings matches the order of the configured AWS regions
+
+**All keyring-producing operations MUST...**
+
+1. Initialize/configure the AWS SDK client(s) required for communicating with AWS KMS for the given key names/AWS regions
+2. Configure each AWS KMS keyring with the AWS SDK client it needs to communicate with AWS KMS
+3. Be configurable with an optional AWS SDK client configuration (client config)
+   or a similar AWS SDK language-specific client configuration option
+   1. If a client config is provided,
+      the keyring-producing operation MUST configure all AWS SDK clients with the provided client config
+4. Limit the initialization of new AWS SDK client(s) when possible
+   1. Multiple keyrings can be configured with the same AWS SDK client if the following conditions are all met:
+      1. Keyrings MUST be configured in the same keyring-producing operation call
+      2. Keyrings MUST communicate with AWS KMS in the same AWS region
+      3. AWS SDK clients MUST share the same AWS SDK client configuration
+5. Maintain the order of all key names/AWS regions provided by the customer
+
+**All keyring-producing operations MAY...**
+
+1. Add an identifier to the user agent string in the AWS SDK client
+   that was initialized by the keyring-producing operation
+   to note the ESDK language and version
+
+**All keyring-producing operations MUST NOT...**
+
+1. Modify any customer-provided client config
+2. Modify the user agent string if the customer has already provided one
```

Comment:
The user agent string is generally a concatenation of separate strings, and the SDKs generally make this an additive process (e.g. https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/RequestClientOptions.html#appendUserAgent-java.lang.String-). Assuming the more important requirement of not modifying parameters, do we need this restriction?

Reply:
Leaving this open. We'll need to have more discussions about this. I do not believe we need to be this restrictive, but this decision may be outside the scope of this PR.